### PR TITLE
Add `Window::has_focus`

### PR DESCRIPTION
**Pull request by amrbashir**
- [X] Tested on all platforms changed
	- [X] Windows
	- [X] X11
	- [ ] Wayland
	- [x] Web 
	- [ ] Android
	- [ ] iOS
	- [x] macOS
- [X] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [X] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [X] Created or updated an example program if it would help users understand this functionality
- [X] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

Porting https://github.com/tauri-apps/tao/commit/7d2eeeebb4da15e9aeda9bf17e80ebdf23c95cee

### CODE REVIEW

File: src/window.rs
Comment (madsmtm) [inline src/window.rs:933]: Nit: Web should probably be "not implemented" instead of "unsupported"

File: src/window.rs
Comment (madsmtm) [inline src/window.rs:928]:
```suggestion
    /// Gets whether the window has keyboard focus.
```

File: src/window.rs
Comment (madsmtm) [inline src/window.rs:929]:
```suggestion
    ///
    /// This queries the same state information as [`WindowEvent::Focused`].
    ///
```

**Review by madsmtm — APPROVED** [review]
Thanks, looks good, only a few minor nits.

**Comment by kchibisov** [comment]
I think it's dangerous to add this with blanket `true` impl to be honest, since users my start relying on it. Given that you simply want to get the last value of `Focused(changed)` event we should have the Api on all the platform implemented from the start.

**Comment by amrbashir** [comment]
Unfortunately, I don't have much experience with other platforms so I can't implement them, however I may be able to implement x11 if you steer me into the right direction.

**Comment by kchibisov** [comment]
@amrbashir I mean it's not about knowing though. You can see where we sent `Fosused` event and stash its result on a window, it's just rust and doesn't require extra knowledge at all. For example on Wayland you'd need to pass some `Arc<AtomicBool>` around, on X11 can likely something similar or maybe you can query things directly.

There're no platform specifics at all in the end.

**Comment by amrbashir** [comment]
cool, thought there might be a way to query the OS like macOS impl for example.

**Comment by kchibisov** [comment]
Maybe on X11, but be aware that active window isn't keyboard focused window sometimes, since you can hover the window with mouse and it'll be active, but won't have focus.

**Comment by amrbashir** [comment]
alright, x11, wayland and web, are implemented now but I was only able to test x11.

File: examples/window.rs
Comment (kchibisov) [inline examples/window.rs:22]: This should be reverted.

File: src/platform_impl/linux/x11/window.rs
Comment (kchibisov) [inline src/platform_impl/linux/x11/window.rs:95]: Shouldn't this be `false`?

File: src/platform_impl/linux/wayland/window/shim.rs
Comment (kchibisov) [inline src/platform_impl/linux/wayland/window/shim.rs:155]:
```suggestion
    /// Whether the window has keyboard focus.
```

**Review by kchibisov — CHANGES_REQUESTED** [review]
While there's a way to know whether the window is focus or not on ios/Android, I guess the issue was that it's unclear how to hook into the current handlers, which is sort of understandable.

Would leave that to android and ios maintainers.

File: src/platform_impl/linux/x11/window.rs
Comment (amrbashir) [reply src/platform_impl/linux/x11/window.rs:95]: doesn't the window start initially with keyboard focus?

**Comment by amrbashir** [comment]
Android and iOS are now implemented, although I am not sure if a global state on Android is the way to go.

**Review by kchibisov — APPROVED** [review]
LGTM.

Would like to hear from android maintainers wrt the change and tracking
focus in a global state. @MarijnS95 or @msiglreith.

Also, you'd need a rebase, I guess.

**Comment by MarijnS95** [comment]
Android _can_ support multiple windows (called `Activity`s) so this _should not_ go into a `static` global.

**Comment by amrbashir** [comment]
> Android can support multiple windows (called `Activitys`) so this should not go into a `static` global.

I am aware of that but glancing over the code, the events are dispatched with a dummy window id and there is no way to differentiate between different activity events.

**Comment by MarijnS95** [comment]
That's unfortunate to hear @amrbashir, I thought @rib's solution in #2444 would implicitly address this; or at least his crate opens the door for multi-window (multi-`Activity` _instances_) setup as that's relevant to some of my use-cases going forward.

Regardless, I thought at least in the past we have a per-window `struct`/storage location where I stored the `RawWindowHandle` lock.
Looks like @rib removed that (probably for good reason, though I have not had the opportunity to look at any of the code in detail yet), and looking back it seems to have been in `EventLoop` which is already a better place than a global `static` (but not per-`Window`, as I remembered), allowing for easier move to a per-window `struct` when we get to implementing this.

File: src/platform_impl/ios/window.rs
Comment (madsmtm) [inline src/platform_impl/ios/window.rs:353]: Nit: using raw `bool` works

**Review by madsmtm — APPROVED** [review]
Haven't tested the iOS impl, but is very trivial, so accepted from my side. macOS also looks good, thanks for your work @amrbashir!

File: src/platform_impl/ios/window.rs
Comment (amrbashir) [reply src/platform_impl/ios/window.rs]: oh, good to know.

File: src/platform_impl/linux/x11/window.rs
Comment (madsmtm) [reply src/platform_impl/linux/x11/window.rs:95]: On macOS, the window is initially not focused (and `window.has_focus()` returns `false`), but it receives a `WindowEvent::Focused` shortly after launching

Comment (amrbashir) [reply src/platform_impl/linux/x11/window.rs:95]: This only affects x11, on macOS we query the OS directly. The reason I made x11 start with `true` is because Windows also starts with `true` and I think x11 is very similar to Windows in this matter, but I haven't had the chance to test x11 (I may test it later today).

**Review by kchibisov — APPROVED** [review]

**Comment by kchibisov** [comment]
@amrbashir could you rebase it?

**Comment by amrbashir** [comment]
Should be fine now

**Pull request merged**
